src/io/flutter/editor/outline/OpenEditorOutlineService.java

```diff
@@ -0,0 +1,232 @@
+        if (listener != null) {
+          analysisServer.removeOutlineListener(path, listener);
+        }
+      }
+
+      // Register new outline listeners.
+      for (final String path : newPaths) {
+        if (outlineListeners.containsKey(path)) continue;
+        final FlutterOutlineListener listener =
+          (filePath, outline, instrumentedCode) -> {
+            synchronized (outlineListeners) {
+              if (!outlineListeners.containsKey(path)) {
+                // The outline listener subscription was already cancelled.
+                return;
+              }
+            }
+            synchronized (pathToOutline) {
+              pathToOutline.put(path, outline);
+            }
+            notifyOutlineUpdated(path);
+          };
+        outlineListeners.put(path, listener);
+        analysisServer.addOutlineListener(FileUtil.toSystemDependentName(path), listener);
+      }
+    }
+
+    synchronized (pathToOutline) {
+      for (final String path : obsoletePaths) {
+        // Clear the current outline as it may become out of date before the
+        // file is visible again.
+        pathToOutline.remove(path);
+      }
+    }
+    if (obsoletePaths.size() > 0 || newPaths.size() > 0) {
+      notifyEditorsUpdated();
+    }
+  }
+
+  private void notifyEditorsUpdated() {
+    for (Listener listener : listeners) {
+      listener.onEditorsChanged();
+    }
+  }
+
+  private void notifyOutlineUpdated(String path) {
+    for (Listener listener : listeners) {
+      listener.onOutlineChanged(path);
+    }
+  }
+
+  public void addListener(@NotNull Listener listener) {
+    listeners.add(listener);
+  }
+
+  public void removeListener(@NotNull Listener listener) {
+    listeners.remove(listener);
+  }
+
+  @Nullable
+  public FlutterOutline get(String path) {
+    return pathToOutline.get(path);
+  }
+
+  @Override
+  public void dispose() {
+    synchronized (outlineListeners) {
+      for (Map.Entry<String, FlutterOutlineListener> entry : outlineListeners.entrySet()) {
+        final String path = entry.getKey();
+        final FlutterOutlineListener listener = outlineListeners.remove(path);
+        if (listener != null) {
+          analysisServer.removeOutlineListener(path, listener);
+        }
+      }
+      outlineListeners.clear();
+    }
+    synchronized (pathToOutline) {
+      pathToOutline.clear();
+    }
+  }
+
+  /**
+   * Listener for changes to the active editors or open outlines.
+   */
+  public interface Listener {
+    /**
+     * Called on a change in the set of open editors.
+     */
+    void onEditorsChanged();
+
+    /**
+     * Called on a change in the outline of file {@param path}.
+     */
+    void onOutlineChanged(String path);
+  }
+
+
+  /**
+   * The cached {@link FlutterOutline} for some file.
+   *
+   * <p>
+   * Provides useful metadata for test identification.
+   */
+  @Immutable
+  public static class OutlineCache {
```

Comment:
super nit: maybe OutlineCacheEntry as OutlineCache makes me think this is an overall cache vs a cache for a single file.